extern/sector-storage/sched_worker.go

```diff
@@ -340,14 +340,15 @@ func (sw *schedWorker) workerCompactWindows() {
 func (sw *schedWorker) processAssignedWindows() {
 	worker := sw.worker
 
-assignLoop:
 	// process windows in order
-	for len(worker.activeWindows) > 0 {
-		firstWindow := worker.activeWindows[0]
+	widx := 0
+	for len(worker.activeWindows) > widx {
+		firstWindow := worker.activeWindows[widx]
```

Comment:
I'd rename `firstWindow` to just `window`, and just use a foreach loop

```suggestion
	for widx, window := range worker.activeWindows {
```

Reply:
Sorry for not watching this for a long time.

This still has a problem, when the first activeWindow has a P2 task and the second activeWindow has a C1 task. But the resources not enough to do P2, but enough to do C1. This code makes C1 wait until P2 is complete.

https://github.com/filecoin-project/lotus/blob/97207df705fc475ead184c14f90c161c8bca5c64/extern/sector-storage/sched_worker.go#L362-L365